### PR TITLE
feat(update): print summary of changes after helm update

### PR DESCRIPTION
**Pull request by adamreese**
Fixes https://github.com/helm/helm/issues/229

### CODE REVIEW

**Comment by adamreese** [comment]
Anyone have feelings on what the summary output should be?  Currently working but ugly

```
$ go run helm.go update
[WARN] A new version of Helm is available. You have 0.1.0-unstable. The latest is 0.1.0
---> Download version 0.1.0 here: https://github.com/helm/helm/releases/tag/0.1.0
---> Checking repo charts
Updated 1 charts
	alpine
Added 19 charts
	cassandra
	docker-registry
	elasticsearch
	example-nginx
	example-todo
	jenkins
	memcached
	mysql
	nginx-alpine
	nginx
	owncloud
	postgresql
	proxy-2-service
	rabbitmq
	redis-cluster
	redis-standalone
	redmine
	riak
	selenium
---> Checking repo arschles
Updated 10 charts
	cassandra
	elasticsearch
	jenkins
	mysql
	owncloud
	postgresql
	rabbitmq
	redis-standalone
	redmine
	riak
---> Done
```

**Comment by technosophos** [comment]
Honestly, I like the full output. I guess you could compress it vertically using a table layout like this: https://github.com/gosuri/uitable

**Comment by technosophos** [comment]
Is this one awaiting review, @adamreese, or are you still working on it?

**Comment by adamreese** [comment]
Awaiting review unless there is more feedback on ui.

File: config/config.go [hunk 1]
```diff
@@ -195,19 +196,86 @@ func ensureRepo(repo, dir string) (*vcs.GitRepo, error) {
...
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", err
+	}
+
+	// cleanup any trailing whitespace
+	return strings.TrimSpace(string(out)), nil
+}
+
+type repoSummary map[string][]string
```
Comment (sgoings) [inline config/config.go]: Should this be closer to the top of the file? /me go-noob

File: config/config.go [hunk 2]
```diff
@@ -195,19 +196,86 @@ func ensureRepo(repo, dir string) (*vcs.GitRepo, error) {
...
+	for _, i := range ignores {
+		if chart == i {
+			return
+		}
+	}
+	s[status] = append(s[status], chart)
+}
+
+// printSummary prints a diff of charts after upate
+func printSummary(diff string) {
```
Comment (sgoings) [inline config/config.go]: Do you _actually_ need this comment (since it's private)?

File: config/config_test.go [hunk 3]
```diff
@@ -73,3 +76,28 @@ func TestSave(t *testing.T) {
...
+M	owncloud`
+
+	expected := "Updated 3 charts\n\tcassandra\n\tmysql\n\towncloud\nAdded 1 charts\n\tjenkins\n"
+	printSummary(diff)
+	actual := b.String()
+
+	if actual != expected {
+		t.Errorf("Expected %q to eq %q", actual, expected)
+	}
+}
```
Comment (sgoings) [inline config/config_test.go]: Yay tests!

**Comment by sgoings** [comment]
I would second the use of uitable if it's not too much extra work. That'll make it more kubectl-output-like as well.

File: config/config.go [hunk 4]
```diff
@@ -195,19 +196,86 @@ func ensureRepo(repo, dir string) (*vcs.GitRepo, error) {
...
 	for _, table := range r.Tables {
-		log.Info("Updating %s", table.Name)
+		log.Info("Checking repo %s", table.Name)
 		rpath := filepath.Join(r.Dir, table.Name)
 		g, err := ensureRepo(table.Repo, rpath)
 		if err != nil {
 			return err
 		}
+
+		initialVersion, _ := g.Version()
```
Comment (arschles) [inline config/config.go]: is the second return val an `error`?

File: config/config.go (hunk 2 above)
Comment (technosophos) [inline config/config.go]: It's considered good practice in Go, though it's not required.

File: config/config.go (hunk 1 above)
Comment (technosophos) [inline config/config.go]: There's no real convention for where you put non-public types, other than "not at the top of the file". The sort of basic idea is that exported (public) stuff floats toward the top, while non-exported stuff is lower down in the file. But if you look at the Go core code, you'll not see this practice followed particularly well.

**Comment by adamreese** [comment]
Still not sure how to format as a table, some repos will have updated charts, some could have added charts.  Putting this in a table would be a lot of wasted real-estate.

```
---> Checking repo charts
Added 1 charts                Updated 0 charts              Deleted 5 charts
Chart1                                                                       ...
---> Done
---> Checking repo periodic
Added 118 charts           Update 0 charts                 Deleted 0 charts
Hydro­gen
He­lium
...
```

@technosophos @sgoings Did you have a different idea for adding to a table?

**Comment by technosophos** [comment]
@adamreese instead of "table", we should have said "columns".

```
⇒  brew update
Updated Homebrew from 68e67acb to 422e608d.
Updated 2 taps (caskroom/cask, homebrew/php).
==> New Formulae
ammonite-repl                codemod                      gdb                          homebrew/php/php54-mongodb   pgroonga                     shairport                    utf8proc
ansible-cmdb                 desk                         generate-json-schema         homebrew/php/php55-gnupg     pldebugger                   shml                         vultr
```

Clearly, this is the sort of thing that we could do on a subsequent PR if you think it's out of scope for this one.

**Comment by adamreese** [comment]
Took a first pass at formatting into columns at https://github.com/adamreese/helm/blob/update-summary-cols/config/config.go#L263-L290

**Comment by technosophos** [comment]
@adamreese nice!

**Comment by adamreese** [comment]
Cherry-picked and squashed.  Can I get another LGTM if it LGTY?

**Pull request merged**